utils/lmtinit.8.in
.TH lmtinit 8 "2022-04-01" "@PACKAGE_NAME@-@PACKAGE_VERSION@" "@PACKAGE_NAME@"
.SH NAME
lmtinit \- initialize LMT database 
.SH SYNOPSIS
.B "lmtinit [OPTIONS]"
.SH DESCRIPTION
.B lmtinit
sets up MySQL for LMT use.
Each file system that will be monitored must have a database
populated with the lmt schema.  This can be set up with the
\fI--add FSNAME\fR option.
.LP
\fBlmtinit\fR requires that the LMT MySQL accounts
(see \fBMYSQL SETUP\fR below) be set up,
and that \fIlmt.conf\fR contains the database host, port, and users with 
appropriate access.
.SH OPTIONS
.B lmtinit 
accepts the following command line options:
.TP
.I "-a,--add FSNAME"
Create a database named \fIfilesystem\_FSNAME\fR and populate it with
tables per the LMT schema.
If database and table creation capabilities are not extended
to the \fIlmt\_db\_rwuser\fR, then a user and password having this
capability must be provided on the command line with
the \fI--user\fR and \fI--password\fR options.
.TP
.I "-d,--delete FSNAME"
Drop a database named \fIfilesystem\_FSNAME\fR.
If database drop capability is not extended
to the \fIlmt\_db\_rwuser\fR, then a user and password having this
capability must be provided on the command line with
the \fI--user\fR and \fI--password\fR options.
.TP
.I "-l,--list"
List the currently configured file systems.
If database list capability is not extended
to the \fIlmt\_db\_rouser\fR, then a user and password having this
capability must be provided on the command line with
the \fI--user\fR and \fI--password\fR options.
.TP
.I "-c,--config-file FILE"
Use an alternate \fIlmt.conf\fR.
.TP
.I "-s,--schema-file FILE"
Use an alternate schema file.
.TP
.I "-u,--user=USER"
Connect to the database as USER.
The default is to use the users configured in \fIlmt.conf\fR.
.TP
.I "-p,--password=PASS"
Connect to the database with password PASS.
The default is to use the passwords configured in \fIlmt.conf\fR.
.TP
.I "-P,--prompt-password"
Same as above but prompt for the password interactively.
.TP
.I "-x,--dump-config"
Print configuration information for read-write database connect.
(Helper function for \flLMT.pm\fR).
.TP
.I "-o,--update-ops FSNAME"
Update the known operation names for FSNAME.
.SH MYSQL SETUP
Refer to MySQL documentation to determine how to set up and secure your
MySQL database, and create the LMT users.
.LP
The \fBmysql_secure_installation\fR program can be used to secure your
database, including changing the database root password.
.LP
LMT needs a read-only and a read/write user for accessing the database.
These users and their passwords are stored in \fIlmt.conf\fR, using
a strategy such as the one shown in \fIlmt.conf(5)\fR for protecting
the read/write password.  Users can be created using a variant of the
\fImkusers.sql\fR example script (see below).
.SH CRON SETUP
The \fBlmt_agg.cron\fR script may optionally be run to create low-resolution
data used to speed up real-time historical graphing in the java client.
See \fIlmt_agg.cron(8)\fR for more information.
.SH OPERATION UPDATING
The operation names (such as read, write, open, etc) that the database
will accept is set at creation time by the schema used.
Newer versions of \fIlmt\fR may have new operation names that older
databases can't accept. Older databases can be updated
to accept the operation names of the current \fIlmt\fR version by using the
\fI--update-ops\fR flag and then restarting \fIcerebrod\fR.
Doing so will not remove any opnames from the database,
so it will continue to be able to accept any opnames that it was
previously able to accept.
.SH FILES
@X_SYSCONFDIR@/@PACKAGE@/lmt.conf
.br
@X_DATADIR@/@PACKAGE@/mkusers.sql
.SH SEE ALSO
lustre(7), lmt.conf(5), mysql_secure_installation(1), lmt_agg.cron(8)


NEWS
========================================================================
Release Notes for LMT 3.2.8                                 25 Nov 2019
========================================================================

* #49 handle lustre version file containing bare version
* #49 Revert "Add /sys in Lustre version checking (#31)"
* Recovery status reported for any status other than COMPLETE
* #33 build: Fix autotools warnings
* #30 Fix "lmtmetric: osc metric: No such file or directory"
* #44 Fully expand ${prefix} in configure so make install works

========================================================================
Release Notes for LMT 3.2.7                                 23 Oct 2019
========================================================================

* #40 ltop fix to run without MDTs or OSTs
* #38 Space and Inodes for MDTs show in ltop

========================================================================
Release Notes for LMT 3.2.6                                 06 May 2019
========================================================================

* #36 look in /sys and /proc as necessary - Lustre 2.12 compatability

========================================================================
Release Notes for LMT 3.2.5                                 17 Sep 2018
========================================================================

* Fix omission of 'z' in ltop.1 man page

========================================================================
Release Notes for LMT 3.2.4                                 17 Sep 2018
========================================================================

* In ltop ensure MDT and OST windows have 3 rows
* ltop monitor file systems with only MDTs or only OSTs

========================================================================
Release Notes for LMT 3.2.3                                 26 Feb 2018
========================================================================

* #29 Add MDT count to FS selection list
* #31 Add /sys in Lustre version checking

========================================================================
Release Notes for LMT 3.2.0                                 01 Apr 2016
========================================================================

* #28 Add MDT sub-window

========================================================================
Release Notes for LMT 3.1.10                                16 Mar 2016
========================================================================

* #27 Report recovery status for MDT

========================================================================
Release Notes for LMT 3.1.9                                 27 Oct 2014
========================================================================

* #24 Fix compilation against Lua 5.2
* #25 Fix _find_all_fs() so it finds all the filesystems

========================================================================
Release Notes for LMT 3.1.8                                 22 Aug 2014
========================================================================

* Update license headers for GPLv2+

* Stop tracking autotools products.

========================================================================
Release Notes for LMT 3.1.6                                 26 Apr 2013
========================================================================

* Ignore missing ost brw_stats file in lustre

========================================================================
Release Notes for LMT 3.1.5                                 03 Apr 2012
========================================================================

* Redraw OST window correctly after selecting new filesystem.

========================================================================
Release Notes for LMT 3.1.4                                 03 Apr 2012
========================================================================

* Issue 13: fix path to mdt stats for Lustre 2.x

* Issue 3: allow switching between filesystems while running

========================================================================
Release Notes for LMT 3.1.3                                 04 Mar 2011
========================================================================

* Issue 53: ltop should not display FQDN in OSS column

========================================================================
Release Notes for LMT 3.1.2                                 11 Feb 2011
========================================================================

* Issue 48: lmtinit cannot handle hyphen in database name

* Issue 50: ltop not working with hyphenated file system name

* ltop: added options description to Usage message.

========================================================================
Release Notes for LMT 3.1.1                                 03 Dec 2010
========================================================================

* Issue 47: ltop should highlight the ost column header used for sorting
  Added a '>' character in front of the current sort field.

========================================================================
Release Notes for LMT 3.1.0                                 03 Dec 2010
========================================================================

* Merge lustre 2.x portability changes from lmt-lustre2x branch.

* Add temporary workaround for missing mdt, ldlm proc files in 2.0.50.zfs
  and also some variant of 1.8.3 discussed in issue 28.

========================================================================
Release Notes for LMT 3.0.4                                 03 Dec 2010
========================================================================

* Issue 45: ltop doesn't show ost's if mdt metric isn't reporting in
  Fix curses update problem in this case

* Issue 46: handle duplicate statfs line in /proc/fs/lustre/mds/target/stats   	
  Change stats hashing code to allow duplicates (last in wins).


utils/lmtinit.c
/*****************************************************************************
 *  Copyright (C) 2010 Lawrence Livermore National Security, LLC.
 *  This module was written by Jim Garlick <garlick@llnl.gov>
 *  UCRL-CODE-232438 All Rights Reserved.
 *
 *  This file is part of the Lustre Monitoring Tool.
 *  For details, see http://github.com/chaos/lmt.
 *
 *  This program is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the license, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF MERCHANTABILITY
 *  or FITNESS FOR A PARTICULAR PURPOSE. See the terms and conditions of the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA or see
 *  http://www.gnu.org/licenses.
 *****************************************************************************/

#if HAVE_CONFIG_H
#include "config.h"
#endif
#include <stdio.h>
#include <stdarg.h>
#include <errno.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>
#include <string.h>
#if HAVE_GETOPT_H
#include <getopt.h>
#endif
#include <libgen.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#include "list.h"
#include "hash.h"
#include "error.h"

#include "util.h"

#include "lmtconf.h"

#include "lmtmysql.h"

#define OPTIONS "a:d:lc:s:u:p:Px:o:"
#if HAVE_GETOPT_LONG
#define GETOPT(ac,av,opt,lopt) getopt_long (ac,av,opt,lopt,NULL)
static const struct option longopts[] = {
    {"add",             required_argument,  0, 'a'},
    {"delete",          required_argument,  0, 'd'},
    {"list",            no_argument,        0, 'l'},
    {"config-file",     required_argument,  0, 'c'},
    {"schema-file",     required_argument,  0, 's'},
    {"user",            required_argument,  0, 'u'},
    {"password",        required_argument,  0, 'p'},
    {"prompt-password", no_argument,        0, 'P'},
    {"dump-config",     no_argument,        0, 'x'},
    {"update-ops",      required_argument,  0, 'o'},
    {0, 0, 0, 0},
};
#else
#define GETOPT(ac,av,opt,lopt) getopt (ac,av,opt)
#endif

#define LMT_SCHEMA_VERSION "1.1"
#define LMT_SCHEMA_PATH \
    X_DATADIR "/" PACKAGE "/create_schema-" LMT_SCHEMA_VERSION ".sql"


static void _list (char *user, char *pass);
static void _del (char *user, char *pass, char *fsname);
static void _add (char *user, char *pass, char *fsname, char *schemafile);
static void _xconf (char *user, char *pass);
static void _update (char *user, char *pass, char *fsname);

static void
usage(void)
{
    fprintf (stderr, "Usage: lmtinit [OPTIONS]\n"
        "  -a,--add FS            create database for file system\n"
        "  -d,--delete FS         remove database for file system\n"
        "  -l,--list              list configured file systems\n"
        "  -c,--config-file FILE  use an alternate config file\n"
        "  -s,--schema-file FILE  use an alternate schema file\n"
        "  -u,--user=USER         connect to the db with USER\n"
        "  -p,--password=PASS     connect to the db with PASS\n"
        "  -P,--prompt-password   prompt for password\n"
        "  -x,--dump-config       dump config in machine readable form\n"
        "  -o,--update-ops FS     update the mdt op names for file system\n"
    );
    exit (1);
}

int
main (int argc, char *argv[])
{
    int c;
    int aopt = 0;
    int dopt = 0;
    int lopt = 0;
    int Popt = 0;
    int xopt = 0;
    int oopt = 0;
    char *fsname = NULL;
    char *conffile = NULL;
    char *schemafile = NULL;
    char *user = NULL;
    char *pass = NULL;

    err_init (argv[0]);
    optind = 0;
    opterr = 0;
    while ((c = GETOPT (argc, argv, OPTIONS, longopts)) != -1) {
        switch (c) {
            case 'a':   /* --add FS */
                aopt = 1;
                fsname = optarg;
                break;
            case 'd':   /* --delete FS */
                dopt = 1;
                fsname = optarg;
                break;
            case 'l':   /* --list */
                lopt = 1;
                break;
            case 'c':   /* --config-file FILE */
                conffile = optarg;
                break;
            case 's':   /* --schema-file FILE */
                schemafile = optarg;
                break;
            case 'u':   /* --user USER */
                user = optarg;
                break;
            case 'p':   /* --password PASS */
                pass = optarg;
                break;
            case 'P':   /* --prompt-password */
                Popt = 1;
                break;
            case 'x':   /* --dump-config */
                xopt = 1;
                break;
            case 'o':  /* --update-ops */
                oopt = 1;
                fsname = optarg;
                break;
            default:
                usage ();
        }
    }
    if (lmt_conf_init (1, conffile) < 0)
        exit (1);
    lmt_conf_set_db_debug (1);
    if (optind < argc)
        usage ();
    if (!aopt && !dopt && !lopt && !xopt && !oopt)
        usage ();
    if (aopt + dopt + lopt + xopt + oopt > 1)
        msg_exit ("Use only one of -a, -d, -l, -o, and -x options.");
    if (pass && Popt)
        msg_exit ("Use only one of -p and -P options.");
    if (xopt && (Popt || user || pass))
        msg_exit ("-x cannot be used with -u, -p, or -P options.");

    if (lopt) {
        if (!user)
            user = lmt_conf_get_db_rouser ();
        if (Popt)
            pass = getpass ("Password: ");
        else if (!pass)
            pass = lmt_conf_get_db_ropasswd ();
    } else {
        if (!user)
            user = lmt_conf_get_db_rwuser ();
        if (Popt)
            pass = getpass ("Password: ");
        else if (!pass)
            pass = lmt_conf_get_db_rwpasswd ();
    }

    if (lopt)
        _list (user, pass);
    else if (dopt)
        _del (user, pass, fsname);
    else if (aopt)
        _add (user, pass, fsname, schemafile);
    else if (xopt)
        _xconf (user, pass);
    else if (oopt)
        _update(user, pass, fsname);

    exit (0);
}

static void
_list (char *user, char *pass)
{
    List l;
    ListIterator itr;
    char *s, *p;

    if (lmt_db_list (user, pass, &l) < 0)
        exit (1);
    itr = list_iterator_create (l);
    while ((s = list_next (itr))) {
        p = strchr (s, '_');
        printf ("%s\n", p ? p + 1 : s);
    }
    list_iterator_destroy (itr);
}

static void
_del (char *user, char *pass, char *fsname)
{
    if (lmt_db_drop (user, pass, fsname) < 0)
        exit (1);
}

static int
_read_schema (char *filename, char **cp)
{
    int fd = -1;
    struct stat sb;
    char *buf = NULL;
    char *path = filename ? filename : LMT_SCHEMA_PATH;
    int n, res = -1;

    if ((fd = open (path, O_RDONLY)) < 0) {
        err ("could not open %s for reading", path);
        goto done;
    }
    if (fstat (fd, &sb) < 0) {
        err ("could not fstat %s", path);
        goto done;
    }
    buf = xmalloc (sb.st_size);
    n = read (fd, buf, sb.st_size);
    if (n < 0) {
        err ("error reading %s", path);
        goto done;
    } else if (n == 0) {
        msg ("error reading %s: premature EOF", path);
        goto done;
    } if (n != sb.st_size) {
        msg ("error reading %s: short read", path);
        goto done;
    }
    res = 0;
    *cp = buf;
done:
    if (fd != -1)
        close (fd);
    if (res < 0)
        free (buf);
    return res;
}

static void
_add (char *user, char *pass, char *fsname, char *schemafile)
{
    char *buf = NULL;
    
    if (_read_schema (schemafile, &buf) < 0)
        exit (1);
    if (lmt_db_add (user, pass, fsname, LMT_SCHEMA_VERSION, buf) < 0)
        exit (1);
    if (buf)
        free (buf);
}

static void
_xconf (char *user, char *pass)
{
    char *host = lmt_conf_get_db_host ();
    int port = lmt_conf_get_db_port ();

    printf ("dbhost:%s\n", host ? host : "");
    printf ("dbport:%d\n", port);
    printf ("dbuser:%s\n", user ? user : "");
    printf ("dbauth:%s\n", pass ? pass : "");
}

static void
_update(char *user, char* pass, char *fsname)
{
    if (lmt_db_update_ops(user, pass, fsname) < 0) {
        printf ("failed to update ops for %s\n", fsname);
    }
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
